File: src/main/java/com/company/erp/core/security/JwtService.java
package com.company.erp.core.security;

import com.company.erp.core.config.AppProperties;
import com.company.erp.core.exceptions.UnauthorizedException;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

/**
 * Issues and parses access / refresh JWTs.
 *
 * <p>Access tokens carry:
 * <ul>
 *   <li>{@code sub}  — user id (Long, string form)</li>
 *   <li>{@code email}</li>
 *   <li>{@code pms}  — permission codes (string array)</li>
 *   <li>{@code typ}  — "access"</li>
 * </ul>
 *
 * <p>Refresh tokens carry only {@code sub}, {@code jti}, and
 * {@code typ=refresh}; their JTI is persisted in {@code refresh_tokens} so
 * they can be revoked / rotated.</p>
 */
@Service
public class JwtService {

    private static final String CLAIM_EMAIL       = "email";
    private static final String CLAIM_PERMISSIONS = "pms";
    private static final String CLAIM_TYPE        = "typ";
    private static final String TYPE_ACCESS       = "access";
    private static final String TYPE_REFRESH      = "refresh";

    private final AppProperties props;
    private final SecretKey key;

    public JwtService(AppProperties props) {
        this.props = props;
        this.key = Keys.hmacShaKeyFor(props.security().jwt().secret().getBytes(StandardCharsets.UTF_8));
    }

    public IssuedToken issueAccess(Long userId, String email, Set<String> permissions) {
        Instant now = Instant.now();
        Instant exp = now.plus(props.security().jwt().accessTokenTtl());
        String jti = UUID.randomUUID().toString();
        String token = Jwts.builder()
                .issuer(props.security().jwt().issuer())
                .subject(userId.toString())
                .id(jti)
                .issuedAt(Date.from(now))
                .expiration(Date.from(exp))
                .claim(CLAIM_EMAIL, email)
                .claim(CLAIM_PERMISSIONS, permissions)
                .claim(CLAIM_TYPE, TYPE_ACCESS)
                .signWith(key)
                .compact();
        return new IssuedToken(token, jti, exp);
    }

    public IssuedToken issueRefresh(Long userId) {
        Instant now = Instant.now();
        Instant exp = now.plus(props.security().jwt().refreshTokenTtl());
        String jti = UUID.randomUUID().toString();
        String token = Jwts.builder()
                .issuer(props.security().jwt().issuer())
                .subject(userId.toString())
                .id(jti)
                .issuedAt(Date.from(now))
                .expiration(Date.from(exp))
                .claim(CLAIM_TYPE, TYPE_REFRESH)
                .signWith(key)
                .compact();
        return new IssuedToken(token, jti, exp);
    }

    public AuthenticatedUser parseAccess(String token) {
        Claims claims = parse(token);
        if (!TYPE_ACCESS.equals(claims.get(CLAIM_TYPE))) {
            throw new UnauthorizedException("Wrong token type");
        }
        Long userId = Long.valueOf(claims.getSubject());
        String email = (String) claims.getOrDefault(CLAIM_EMAIL, "");
        @SuppressWarnings("unchecked")
        List<String> perms = (List<String>) claims.get(CLAIM_PERMISSIONS);
        Set<String> permissions = perms == null ? Set.of() : new HashSet<>(perms);
        return new AuthenticatedUser(userId, email, permissions);
    }

    public RefreshClaims parseRefresh(String token) {
        Claims claims = parse(token);
        if (!TYPE_REFRESH.equals(claims.get(CLAIM_TYPE))) {
            throw new UnauthorizedException("Wrong token type");
        }
        return new RefreshClaims(
                Long.valueOf(claims.getSubject()),
                claims.getId(),
                claims.getExpiration().toInstant());
    }

    private Claims parse(String token) {
        try {
            return Jwts.parser()
                    .verifyWith(key)
                    .requireIssuer(props.security().jwt().issuer())
                    .build()
                    .parseSignedClaims(token)
                    .getPayload();
        } catch (JwtException ex) {
            throw new UnauthorizedException("Invalid or expired token");
        }
    }

    public record IssuedToken(String value, String jti, Instant expiresAt) {}
    public record RefreshClaims(Long userId, String jti, Instant expiresAt) {}
}

File: app/Support/Auth/JwtService.php
<?php

namespace App\Support\Auth;

use App\Support\Exceptions\UnauthorizedException;
use Illuminate\Support\Str;

/**
 * Issues and parses access / refresh JWTs (HMAC-SHA256), a faithful port of
 * the Spring JwtService.
 *
 * Access tokens carry: sub (user id), email, pms (permission codes), typ=access.
 * Refresh tokens carry: sub, jti, typ=refresh — the jti is persisted in
 * refresh_tokens so the token can be rotated / revoked.
 */
class JwtService
{
    const ALG = 'HS256';
    const CLAIM_EMAIL = 'email';
    const CLAIM_PERMISSIONS = 'pms';
    const CLAIM_TYPE = 'typ';
    const TYPE_ACCESS = 'access';
    const TYPE_REFRESH = 'refresh';

    /** @var string */
    private $issuer;

    /** @var string */
    private $secret;

    /** @var int seconds */
    private $accessTtl;

    /** @var int seconds */
    private $refreshTtl;

    public function __construct()
    {
        $this->issuer = (string) config('erp.security.jwt.issuer');
        $this->secret = (string) config('erp.security.jwt.secret');
        $this->accessTtl = (int) config('erp.security.jwt.access_token_ttl');
        $this->refreshTtl = (int) config('erp.security.jwt.refresh_token_ttl');
    }

    /**
     * @param array<int, string> $permissions
     * @return array{value:string, jti:string, expiresAt:int}
     */
    public function issueAccess(int $userId, string $email, array $permissions): array
    {
        $now = time();
        $exp = $now + $this->accessTtl;
        $jti = (string) Str::uuid();
        $token = Jwt::encodeHs256([
            'iss' => $this->issuer,
            'sub' => (string) $userId,
            'jti' => $jti,
            'iat' => $now,
            'exp' => $exp,
            self::CLAIM_EMAIL => $email,
            self::CLAIM_PERMISSIONS => array_values($permissions),
            self::CLAIM_TYPE => self::TYPE_ACCESS,
        ], $this->secret);

        return ['value' => $token, 'jti' => $jti, 'expiresAt' => $exp];
    }

    /**
     * @return array{value:string, jti:string, expiresAt:int}
     */
    public function issueRefresh(int $userId): array
    {
        $now = time();
        $exp = $now + $this->refreshTtl;
        $jti = (string) Str::uuid();
        $token = Jwt::encodeHs256([
            'iss' => $this->issuer,
            'sub' => (string) $userId,
            'jti' => $jti,
            'iat' => $now,
            'exp' => $exp,
            self::CLAIM_TYPE => self::TYPE_REFRESH,
        ], $this->secret);

        return ['value' => $token, 'jti' => $jti, 'expiresAt' => $exp];
    }

    public function parseAccess(string $token): AuthenticatedUser
    {
        $claims = $this->parse($token);
        if (($claims['typ'] ?? null) !== self::TYPE_ACCESS) {
            throw new UnauthorizedException('Wrong token type');
        }
        $email = isset($claims[self::CLAIM_EMAIL]) ? (string) $claims[self::CLAIM_EMAIL] : '';
        $perms = $claims[self::CLAIM_PERMISSIONS] ?? [];
        $permissions = is_array($perms) ? array_map('strval', $perms) : [];

        return new AuthenticatedUser((int) $claims['sub'], $email, $permissions);
    }

    /**
     * @return array{userId:int, jti:string, expiresAt:int}
     */
    public function parseRefresh(string $token): array
    {
        $claims = $this->parse($token);
        if (($claims['typ'] ?? null) !== self::TYPE_REFRESH) {
            throw new UnauthorizedException('Wrong token type');
        }

        return [
            'userId' => (int) $claims['sub'],
            'jti' => (string) ($claims['jti'] ?? ''),
            'expiresAt' => (int) ($claims['exp'] ?? 0),
        ];
    }

    /**
     * @return array<string, mixed>
     */
    private function parse(string $token): array
    {
        try {
            $decoded = Jwt::decodeHs256($token, $this->secret);
        } catch (\Throwable $e) {
            throw new UnauthorizedException('Invalid or expired token');
        }

        if (($decoded['iss'] ?? null) !== $this->issuer) {
            throw new UnauthorizedException('Invalid or expired token');
        }

        return $decoded;
    }
}
